# TALLER DE OPTIMIZACIÓN

### Presentado por: Daniel Diaz - Alejandro Suarez


Punto 1:
-----------

Considere el siguiente problema, donde el valor de $k$ todavía no se ha establecido,

max $Z = x_{1} + 2x_{2}$

s.a.

$-x_{1} + x_{2} \leq 2 $

$x_{2} \leq 3$

$k x_{1} + x_{2} \leq 2k + 3$

$x_{1} \geq 0, x_{2} \geq 0$

La solución que se usa por ahora es $x_{1}=2$, $x_{2}=3$. Utilice el análisis gráfico para determinar los valores de k tales que esta solución sea de hecho óptima.

<img src="files/punto1.png">

Punto 2:
-----------

max $Z = -x_{1} + 4x_{2}$

s.a.

$-3x_{1} + x_{2} \leq 6$

$x_{1} + 2x_{2} \leq 3$

$x_{2} \geq -3$

$x_{1}$ irrestricta

a) Resuelva este problema usando el método gráfico.

<img src="files/punto2.png">

b)Reformule este problema de manera que tenga solo dos restricciones funcionales y todas las variables tengan restricciones de no negatividad

$x_{1} = x_{1}^{'} - x_{1}^{''}$

$x_{2} = x_{2}^{'} - 3$

c) Aplique el método Simplex paso a paso para resolver este problema.

In [1]:
from scipy.optimize import linprog
import numpy as np


#objective fuction
z = np.array([-1,4])

#Constraints
C = np.array([
    [-3,1],          #C1
    [1,2],
    [0,-1]           #C2
])

#upper_bound
b = np.array([6,3,3])

x1 = (None, None)
x2 = (None, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

OF = z*sol.x
print("z = {}".format(OF.sum()))


     fun: -9.857142857142858
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([0.        , 0.        , 5.14285714])
  status: 0
 success: True
       x: array([-1.28571429,  2.14285714])
x1 = -1.2857142857142858, x2 = 2.1428571428571432, z = 9.857142857142858
z = 9.85714285714286


Punto 3:
-----------

Utilice la idea fundamental del método Simplex para identificar los números que faltan en la tabla:

|     | x1 | x2 | x3 | x4 | x5 | x6 |  RHS |
|-----|----|----|----|----|----|----|:----:|
|  z  |  0 | 7  | 0  | 2  | 0  | 2  | 6    |
| x5  | 0  | 4  | 0  | 1  | 1  | 2  | 7    |
|  x3 | 0  | 4  | 1  | -2 | 0  | 4  | 0    |
| x1  | 1  | 0  | 0  | 1  | 0  | -1 | 1    |

Usando:

|   Z  | x0 | XN | RHS |
|-----|----|----|:----:|
|  1  |  0 | cB^(-1) * B ^ -1 * N - cN ^ t | cB^(-1) * B ^ -1  |
| 0  | I  |  B ^ -1 * N | B ^ -1 * b  |


Dónde:

- B^(-1) * N:

|  	|  	|  	|
|----	|---	|----	|
|  1 	| 1 	| 2 	|
| -2 	| 0 	| 4 	|
| 1 	| 0 	| -1 	|

- N (Coeficientes de las restricciones):

|  	|  	|  	|
|----	|---	|----	|
|  2 	| 2 	| 1/2 	|
| -4 	| -2 	| -3/2	|
| 1 	| 2	| 1/2 	|


- Cb ^ t: Coeficientes de las variables básicas.

| X5 	| X3 	|  X1	|
|----	|---	|----	|
|  0 	| 2 	| 6	|

Punto 4:
-----------
Resolver por simplex:

A-)

In [2]:
#Funcion objetivo
z = np.array([2,-1,1])

#Restricciones
C = np.array([
    [2,1,-2],       
    [-4,1,-2],
    [-2,-3,1]       
])

#Limites superiores
b = np.array([8,-2,-4])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2, x3), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

     fun: -4.0
 message: 'Optimization failed. The problem appears to be unbounded.'
     nit: 3
   slack: array([4., 6., 0.])
  status: 3
 success: False
       x: array([2., 0., 0.])
x1 = 2.0, x2 = 0.0, z = 4.0


El problema es no acotado

B-)


In [3]:
#Funcion objetivo
z = np.array([1,-8])

#Restricciones
C = np.array([
    [-1, -1], 
    [-1, 6],
    [0, 1] 
])

#Limites superiores
b = np.array([-1, 3, 2])

#No negatividad
x1 = (0, None)
x2 = (0, None)

sol = linprog(z, A_ub = C, b_ub = b, bounds = (x1, x2), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

     fun: -6.999999999999998
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([10.,  0.,  0.])
  status: 0
 success: True
       x: array([9., 2.])
x1 = 9.000000000000002, x2 = 1.9999999999999998, z = 6.999999999999998


Aproximando un poco,
$x_{1} = 9, x_{2}=2, z = 7$

C-)


In [4]:
#Funcion objetivo
z = np.array([1,3,-1])

#Restricciones
C = np.array([
    [-2,-1,-1],          #C1
    [1,-1,0],
    [-1,5,1]           #C2
])

#Limite superior
b = np.array([-3,-2,4])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)

sol = linprog(z, A_ub = C, b_ub = b, bounds = (x1, x2, x3), method='simplex')

print(sol)


     fun: 1.75
 message: 'Optimization failed. Unable to find a feasible starting point.'
     nit: 3
  status: 2
 success: False
       x: nan


No tiene solución

D-)

In [5]:
#Funcion objetivo 
z = np.array([4,5,-3])

#Restricciones
C = np.array([
    [1,1,1],          
    [-1,-1,-1],
    [-1,1,0],
    [1,3,1]          
])

#Limites superiores
b = np.array([-10,-10,-1,14])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2, x3), method='simplex')

print(sol)


     fun: 20.0
 message: 'Optimization failed. Unable to find a feasible starting point.'
     nit: 1
  status: 2
 success: False
       x: nan


No tiene solución


Punto 5:
--------

a-) Resuelva gráficamente

<img src="files/punto5.png">

b-) Monte el cuadro inicial para el algoritmo Simplex

Reemplazando $X1 = -X'_{1}$

max $Z'=X'_{1}+4X_{2}$

Para el cuadro:

$-X'_{1}-4X_{2} = 0$

$3X'_{1} + X_{2} + X_{3} = 6$

$-X'_{1}+2X_{2}+X_{4}=4$

$X'_{1}+X_{5}=2$

|    | x1 | x2 | x3 | x4 | x5 | RHS |
|----|----|----|----|----|----|:---:|
| z  | -1 | -4 | 0  | 0  | 0  | 0   |
| x3 | 3  | 1  | 1  | 0  | 0  | 6   |
| x4 | -1 | 2  | 0  | 1  | 0  | 4   |
| x5 | 1  | 0  | 0  | 0  | 1  | 2   |

C-)
<img src="files/excelc.png">

Reemplazando $X_{1}$

$X_{1}=-1.132857$

$X_{2}=2.571428$

$Z = 11.4285714$

Punto 6:
-----
a-) 
Sustituyendo

$X_{2} = -X'_{2}$

$X_{3} = X'_{3} - 10$

Max $Z'=-X_{1}-2X'_{2}+X'_{3}-10$

|    | x1 | x'2| x'3| x4 | x5 | RHS |
|----|----|----|----|----|----|:---:|
| z  | 1  | 2  | -1 | 0  | 0  | 0   |
| x4 | 0  | -3 | 1  | 1  | 0  | 130 |
| x5 | 1  |  1 | -4 | 0  | 1  | 40  |

B-)

<img src="files/exceld.png">

No se ha llegado al óptimo y ninguna variable puede salir de la base por lo que no tiene solución.


Punto 7:
-------


In [10]:
#Funcion objetivo
z = np.array([6,3,4])

#Restricciones
C = np.array([
    [1,6,1],          
    [-1,-6,-1],
    [2,3,1],
    [-2,-3,-1],

])

#Limites superiores
b = np.array([10,-10,15,-15])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)

sol = linprog(z, A_ub = C, b_ub = b, bounds = (x1, x2,x3), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

OF = z*sol.x
print("z = {}".format(OF.sum()))



     fun: 41.66666666666667
 message: 'Optimization terminated successfully.'
     nit: 5
   slack: array([ 0.,  0., -0.,  0.])
  status: 0
 success: True
       x: array([6.66666667, 0.55555556, 0.        ])
x1 = 6.666666666666667, x2 = 0.5555555555555556, z = -41.66666666666667
z = 41.666666666666664


In [7]:

#Funcion objetivo
z = np.array([-8,3,-5])

#Restricciones
C = np.array([
    [1,2,3],          
    [-1,-2,-3],
    [-2,1,-8],
])

#Limite superior
b = np.array([20,-20,-30])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2,x3), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

OF = z*sol.x
print("z = {}".format(OF.sum()))


     fun: 10.0
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([ 0., -0.,  0.])
  status: 0
 success: True
       x: array([0.        , 3.68421053, 4.21052632])
x1 = 0.0, x2 = 3.6842105263157894, z = -10.0
z = -9.999999999999998


Punto 8:
-----

a-) Grafica
<img src="files/punto8a.png">


$X_{1} = 0.86$

$X_{2} = 1.43$

$Z = 3.428$

Problema dual

min $w = 6y_{1} - 2y_{2}$

sujeta a:

$2y_{1} + y_{2} \geq -1$

$3y_{1} - 2y_{2} \geq 3$

$y_{1} \geq 0$

$y_{2} \leq 0$

Graficamente
<img src="files/punto8b.png">

$y_{1} = 0.14$

$y_{2} = -1.29$

$w = 3.43$

b -) Grafica
No es posible resolverlo graficamente

Problema dual

max $w = 2y_{1} - 2y_{2}$

Sujeto a:

$y_{1} - 2y_{2} \leq 2$

$6y_{1} + 5y_{2} \leq 15$

$3y_{1} - y_{2} \leq 5$

$y_{1} + 3y_{2} \leq 6$

$ y_{1}, y_{2} \leq 0$

Graficamente 
<img src="files/punto8bb.png">

$y_{1} = 0$

$y_{2} = -1$

$w = 2$

Punto 9:
-------

Condición de factibilidad dial:

$x_{1} = -x'_{1}$

Así:

Max $z = -2x'_{1} - 3x_{2}$

Sujeta a:

$x'_{1} - x_{2} \leq -3$

$-3x'_{1} + x_{2} \leq 6$


In [8]:

#Funcion objetivo
z = np.array([2,3])

#Restricciones
C = np.array([
    [1,-1],          
    [-3,1],
])

#Limites superiores
b = np.array([-3,6])

#No negatividad
x1 = (0, None)
x2 = (0, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))

OF = z*sol.x
print("z = {}".format(OF.sum()))

     fun: -9.0
 message: 'Optimization failed. The problem appears to be unbounded.'
     nit: 1
   slack: array([3., 0.])
  status: 3
 success: False
       x: array([0., 3.])
x1 = 0.0, x2 = 3.0, z = 9.0
z = 9.0


In [9]:

#Funcion objetivo
z = np.array([-2,-3,-5,-6])

#Restricciones
C = np.array([
    [-1,-2,-3,-1],          
    [-1,1,-1,3],
])

#Limites superiores
b = np.array([-2,-3])

#No negatividad
x1 = (0, None)
x2 = (0, None)
x3 = (0, None)
x4 = (0, None)

sol = linprog(-z, A_ub = C, b_ub = b, bounds = (x1, x2, x3, x4), method='simplex')

print(sol)

print("x1 = {}, x2 = {}, z = {}".format(sol.x[0], sol.x[1], sol.fun*-1))
#Se multiplica por -1 al ser una minimización
OF = z*sol.x
print("z = {}".format(OF.sum() *-1))

     fun: 6.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([1., 0.])
  status: 0
 success: True
       x: array([3., 0., 0., 0.])
x1 = 3.0, x2 = 0.0, z = -6.0
z = 6.0
